x-pack/plugins/infra/public/components/log_stream/index.tsx

```diff
@@ -44,7 +44,7 @@ export const LogStream: React.FC<LogStreamProps> = ({
     isLoadingSourceConfiguration,
   } = useLogSource({
     sourceId,
-    fetch: services.http.fetch,
+    fetch: services?.http?.fetch!,
```

Comment:
`services` should always be there, http may or may not be. We should bail before we call `useLogSource` and throw an error `if (!services.http)`, I think.

Reply:
https://github.com/elastic/kibana/blob/master/src/plugins/kibana_react/public/context/context.tsx#L28

Reply:
I thought about that but I wasn't sure it would work with the rules of hooks.

Reply:
Yeah hmm. I don't like the idea of the component just blowing up because someone removes it from inside of the right provider, with a hard to track error. Are those hook rules "per-component"? Maybe something like this would work?
```tsx
const LogStream = props => {
  const { services } = useKibana();
  if (!services.http) {
    throw new Error('descriptive message');
  }
  return <SafeLogStream fetch={services.http.fetch} {...props} />;
}
```

Reply:
component names TBD 😆

Reply:
With this pattern, we could accommodate people who want to do injected core deps by exporting the "SafeLogStream" component at any time, and we can provide constructive feedback for people who accidentally break the shared component by removing it from being inside the right context...

Reply:
I tried and throwing an exception doesn't break the rules of hooks, so I can do it safely. The error will look like this:

<img width="1281" alt="Screenshot 2020-09-02 at 10 26 16" src="https://user-images.githubusercontent.com/57448/91960443-34d43600-ed0a-11ea-8628-fa525630df35.png">

Unless other teams _really_ need it and push for it, I would prefer to not offer a escape hatch from using `kibana-react`'s providers... I want to encourage people to use the accepted conventions while reducing our maintenance burden.

Reply:
LGTM. I don't think it's a big problem to allow DI eventually, because it's not really an escape hatch as much as the context version is a convenience, but I like starting with this and if nobody ever needs the DI version, even better.